### PR TITLE
ofi nic selection bugfix and document update

### PR DESCRIPTION
**Pull request by wenduwan**
Related to https://github.com/open-mpi/ompi/pull/12135. I got tips from Jeff that the NIC selection logic has changed a lot from 4.x, and the community could use a better document to understand the issue given the many conversations recently.

I went ahead and revisited the code. Indeed we should update the documentation - while doing that I also realized that the new code is not doing the desired thing with the round-robin selection and that is a bug I introduced in the other PR:
- Unbound processes, i.e. `--bind-to none`, always choose the first NIC. VERY BAD!
- The selection process does not filter out NICs/providers of different types from the first provider. That is a latent bug causing processes to use different types of NICs, also VERY BAD!

### CODE REVIEW

File: opal/mca/common/ofi/common_ofi.h [hunk 1]
```diff
@@ -103,47 +103,47 @@ OPAL_DECLSPEC int opal_common_ofi_is_in_list(char **list, char *item);
...
- *                             provider NIC. The provider name and
- *                             attributes are used to restrict NIC
- *                             selection. This provider is returned if the
- *                             NIC selection fails.
- *
- * @param provider (OUT)       struct fi_info* object with the selected
- *                             provider if the selection succeeds
- *                             if the selection fails, returns the fi_info
- *                             object that was initially provided.
+ * The selection is based on the following priority:
```
Comment (wenduwan) [inline opal/mca/common/ofi/common_ofi.h:106]: @jsquyres Does this explain the NIC selection well enough? What else do you think will make it better?

**Review by hppritcha — APPROVED** [review]

**Comment by wenduwan** [comment]
Hey @jsquyres could you give the change a read through and see if the document around nic selection makes sense to you?

**Comment by wenduwan** [comment]
@jsquyres Since this PR contains a bugfix, I will merge this change first. We can sync up later and address your concerns separately.

**Comment by jsquyres** [comment]
@wenduwan So sorry for the delay.  Yes, I agree, the comments describe how the code works.  You were correct to merge.

**Pull request merged**